File: dotnet/test/Microsoft.AutoGen.Core.Tests/TestAgent.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// TestAgent.cs

using System.Text.Json;
using Microsoft.AutoGen.Contracts;
using Microsoft.Extensions.Logging;

namespace Microsoft.AutoGen.Core.Tests;

public class TextMessage
{
    public string Source { get; set; } = "";
    public string Content { get; set; } = "";
}

public class RpcTextMessage
{
    public string Source { get; set; } = "";
    public string Content { get; set; } = "";
}

/// <summary>
/// The test agent is a simple agent that is used for testing purposes.
/// </summary>
public class TestAgent(AgentId id,
        IAgentRuntime runtime,
        Logger<BaseAgent>? logger = null) : BaseAgent(id, runtime, "Test Agent", logger),
        IHandle<TextMessage>,
        IHandle<string>,
        IHandle<RpcTextMessage, string>

{
    public ValueTask HandleAsync(TextMessage item, MessageContext messageContext)
    {
        ReceivedMessages[item.Source] = item.Content;
        return ValueTask.CompletedTask;
    }

    public ValueTask HandleAsync(string item, MessageContext messageContext)
    {
        ReceivedItems.Add(item);
        return ValueTask.CompletedTask;
    }

    public ValueTask HandleAsync(int item, MessageContext messageContext)
    {
        ReceivedItems.Add(item);
        return ValueTask.CompletedTask;
    }

    public ValueTask<string> HandleAsync(RpcTextMessage item, MessageContext messageContext)
    {
        ReceivedMessages[item.Source] = item.Content;
        return ValueTask.FromResult(item.Content);
    }

    public List<object> ReceivedItems { get; private set; } = [];

    /// <summary>
    /// Key: source
    /// Value: message
    /// </summary>
    protected Dictionary<string, object> _receivedMessages = new();
    public Dictionary<string, object> ReceivedMessages => _receivedMessages;
}

[TypeSubscription("TestTopic")]
public class SubscribedAgent : TestAgent
{
    public SubscribedAgent(AgentId id,
        IAgentRuntime runtime,
        Logger<BaseAgent>? logger = null) : base(id, runtime, logger)
    {
    }
}

[TypeSubscription("TestTopic")]
public class SubscribedSaveLoadAgent : TestAgent
{
    public SubscribedSaveLoadAgent(AgentId id,
        IAgentRuntime runtime,
        Logger<BaseAgent>? logger = null) : base(id, runtime, logger)
    {
    }

    public override ValueTask<JsonElement> SaveStateAsync()
    {
        var jsonDoc = JsonSerializer.SerializeToElement(_receivedMessages);
        return ValueTask.FromResult(jsonDoc);
    }

    public override ValueTask LoadStateAsync(JsonElement state)
    {
        _receivedMessages = JsonSerializer.Deserialize<Dictionary<string, object>>(state) ?? throw new InvalidOperationException("Failed to deserialize state");
        return ValueTask.CompletedTask;
    }
}

/// <summary>
/// The test agent showing an agent that subscribes to itself.
/// </summary>
[TypeSubscription("TestTopic")]
public class SubscribedSelfPublishAgent(AgentId id,
        IAgentRuntime runtime,
        Logger<BaseAgent>? logger = null) : BaseAgent(id, runtime, "Test Agent", logger),
        IHandle<string>,
        IHandle<TextMessage>
{
    public async ValueTask HandleAsync(string item, MessageContext messageContext)
    {
        TextMessage strToText = new TextMessage
        {
            Source = "TestTopic",
            Content = item
        };
        // This will publish the new message type which will be handled by the TextMessage handler
        await this.PublishMessageAsync(strToText, new TopicId("TestTopic"));
    }
    public ValueTask HandleAsync(TextMessage item, MessageContext messageContext)
    {
        _text = item;
        return ValueTask.CompletedTask;
    }

    private TextMessage _text = new TextMessage { Source = "DefaultTopic", Content = "DefaultContent" };
    public TextMessage Text => _text;
}

File: typescript/test/core/helpers/TestAgent.ts
import { BaseAgent } from "../../../src/core/BaseAgent";
import { IHandle } from "../../../src/contracts/IHandle";
import { AgentId, IAgentRuntime, TopicId } from "../../../src/contracts/IAgentRuntime"; // Add TopicId
import { MessageContext } from "../../../src/contracts/MessageContext";
import { TypeSubscription } from "../../../src/core/TypeSubscriptionAttribute";

export interface TextMessage {
  source: string;
  content: string;
}

export interface RpcTextMessage {
  source: string;
  content: string;
}

export class TestAgent extends BaseAgent implements IHandle<TextMessage>, IHandle<string>, IHandle<RpcTextMessage> {
  public receivedItems: unknown[] = [];
  protected receivedMessages: Map<string, unknown> = new Map();

  constructor(id: AgentId, runtime: IAgentRuntime) {
    super(id, runtime, "Test Agent");
  }

  async handleAsync(message: TextMessage | string | RpcTextMessage, context: MessageContext): Promise<unknown> {
    console.log(`TestAgent.handleAsync:`, {
      message,
      context,
      agentId: this.id,
      messageType: typeof message
    });
    
    if (typeof message === "string") {
      console.log('Handling string message');
      this.receivedItems.push(message);
      return message;  // Return the string message for RPC
    }

    if ('source' in message && 'content' in message) {
      console.log(`Handling TextMessage: ${message.source} - ${message.content}`);
      this.receivedMessages.set(message.source, message.content);
      return message.content;  // Return content for RPC
    }
    console.log('Unknown message type');
    return null;
  }

  get ReceivedMessages(): Record<string, unknown> {
    return Object.fromEntries(this.receivedMessages);
  }
}

@TypeSubscription("TestTopic")
export class SubscribedAgent extends TestAgent {
  constructor(id: AgentId, runtime: IAgentRuntime) {
    super(id, runtime);  // Remove "Test Agent" as it's handled in TestAgent
  }
}

@TypeSubscription("TestTopic")
export class SubscribedSaveLoadAgent extends TestAgent {
  constructor(id: AgentId, runtime: IAgentRuntime) {
    super(id, runtime);  // Remove "Test Agent" as it's handled in TestAgent
  }

  async saveStateAsync(): Promise<unknown> {
    return this.ReceivedMessages;
  }

  async loadStateAsync(state: unknown): Promise<void> {
    if (typeof state === "object" && state !== null) {
      this.receivedMessages = new Map(Object.entries(state));
    }
  }
}

@TypeSubscription("TestTopic")
export class SubscribedSelfPublishAgent extends BaseAgent implements IHandle<string>, IHandle<TextMessage> {
  private _text: TextMessage = { source: "DefaultTopic", content: "DefaultContent" };

  constructor(id: AgentId, runtime: IAgentRuntime) {
    super(id, runtime, "Test Self-Publishing Agent");
  }

  // Combine the two handleAsync methods using type guards
  async handleAsync(item: string | TextMessage, messageContext: MessageContext): Promise<void> {
    if (typeof item === 'string') {
      const strToText: TextMessage = {
        source: "TestTopic",
        content: item
      };
      await this.runtime.publishMessageAsync(strToText, { type: "TestTopic", source: "test" }, this.id);
    } else {
      this._text = item;
    }
  }

  get Text(): TextMessage {
    return this._text;
  }
}
